### PR TITLE
bpo-30345: Add -g to LDFLAGS to ease debug

### PR DESCRIPTION
**Pull request by vstinner**
When using PGO+LTO and compile without LDFLAGS=-g, gdb fails to get
function arguments: add -g to LDFLAGS to ease debug.

https://bugs.python.org/issue30345

### CODE REVIEW

**Comment by vstinner** [comment]
cc  @Traceur759

File: configure [hunk 1]
```diff
@@ -6716,6 +6716,9 @@ then
 	    else
 		OPT="-g $WRAP -O3 -Wall"
 	    fi
+	    # bpo-30345: When using PGO+LTO and compile without LDFLAGS=-g,
+	    # gdb fails to get function arguments.
+	    LDFLAGS="$LDFLAGS -g"
```
Comment (serhiy-storchaka) [inline configure]: Shouldn't it be added only in debug mode? I afraid that adding `-g` in release mode can disable some optimizations.

**Review by methane — APPROVED** [review]

File: configure (hunk 1 above)
Comment (vstinner) [reply configure]: We already *always* pass -g to CFLAGS, even in release mode.

Extract of gcc manual page:

```
       -g  Produce debugging information in the operating system's native
           format (stabs, COFF, XCOFF, or DWARF).  GDB can work with this
           debugging information.

           On most systems that use stabs format, -g enables use of extra
           debugging information that only GDB can use; this extra information
           makes debugging work better in GDB but probably makes other
           debuggers crash or refuse to read the program.  If you want to
           control for certain whether to generate the extra information, use
           -gstabs+, -gstabs, -gxcoff+, -gxcoff, or -gvms (see below).
```

It's unrelated to optimizations. It's just an option to ask to include also debug symbols in the produced binary file (like ELF file on Linux).

If you want to get ride of them, just run "strip python". Linux vendors are able to distribute these debug symbols in a different package, like python-debuginfo on Fedora.

I'm not 100% sure, that's why I'm asking for a review :-)

**Comment by pitrou** [comment]
I think it's ok to add `-g` to link flags, I just don't understand why it's added in a single code path in `configure.ac`.
(also, what exactly is `$ac_cv_prog_cc_g`?)

**Comment by methane** [comment]
AFAIK, `-g` is required only when LTO because code generation is happen in link time.
So I think it's safer that adding `-g` here in configure.ac.

```
   CFLAGS="$CFLAGS $LTOFLAGS"
-  LDFLAGS="$LDFLAGS $LTOFLAGS"
+  LDFLAGS="$LDFLAGS -g $LTOFLAGS"
```

LINKCC is CC for most case, but not always.

**Comment by methane** [comment]
> (also, what exactly is $ac_cv_prog_cc_g?)

I think it means whether CC supports `-g` or not.  And we use CC for LINKCC by default on most platform.  (except AIX).

**Comment by vstinner** [comment]
"AFAIK, -g is required only when LTO because code generation is happen in link time. So I think it's safer that adding -g here in configure.ac."

Done.

**Comment by vstinner** [comment]
"(also, what exactly is $ac_cv_prog_cc_g?) I think it means whether CC supports -g or not."

Hum, why, I cannot add -g for all platforms?

**Comment by methane** [comment]
I don't know linker without -g support.
If AIX supports -g, I feel worrying about minor linker/platform is worthless.

**Comment by vstinner** [comment]
I modified my PR to respect $ac_cv_prog_cc_g.

**Comment by methane** [comment]
I'm sorry, I meant why ac_cv_prog_cc_g  is used in general.

LTO is used only with clang and gcc.  So no need to worry about others when adding `-g` only with LTO.

**Comment by methane** [comment]
I'm OK to merge this with or without $ac_cv_prog_cc_g check.

File: configure (hunk 1 above)
Comment (serhiy-storchaka) [reply configure]: Thank you, now I got this.

**Pull request merged**